kvserver: annotate queues with profiler labels

wenyihu6 opened the pull request:
Prior to this commit, the debugging process was complicated because it was
difficult to determine which specific queue invoked a particular function in the
baseQueue while examining the goroutine dump. This lack of clarity is because
different types of queues (consistency, merge, replica, etc.) all utilize the
baseQueue as their base implementation.

To resolve this issue, this commit adds pprof labels to key baseQueue functions
like maybeAdd and processLoop. As a result, when examining the goroutine dump,
the pprof label (queueName, "") now displays the queue name if the queue is
stuck on the function call.

Fixes: https://github.com/cockroachdb/cockroach/issues/102222
Release note: None

<img width="857" alt="Screenshot 2023-05-17 at 11 18 54" src="https://github.com/cockroachdb/cockroach/assets/56973754/3646bf7b-8754-4d5f-b18d-3efcb9f7451d">

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103525)
<!-- Reviewable:end -->

kvoli reviewed (commented):
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103525)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NVjWmAw4CPb6IABzNPD):*
Those queues still re-use the baseQueue as the base implementation for their queue after this commit. Perhaps try rewording the follow on sentence and using that after "Prior to this commit".
___
*[`pkg/kv/kvserver/queue.go` line 698 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NVjYH2U79JflwE4j-a0:-NVjYH2V1L7ZJZnfcFYD:b-ovjitk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dedabe181b7a7735aed2cd1e86144f8a50188bbd/pkg/kv/kvserver/queue.go#L698)):*
> ```Go
> 	realRepl, _ := repl.(*Replica)
> 
> 	labels := pprof.Labels("queue name", bq.name)
> ```

It would be good to capture the above baseQueue logic with the pprof labels, have you tried shifting up the pprof labeling just after we annotate the ctx  above? 

Does this approach show only `{queue_name: "..."}`? I think the label here is also different than the one below on `processLoop`, which would not include the key `queue_name` e.g. `{:"..."}`. 

A benefit of using ` pprofutil.SetProfilerLabelsFromCtxTags(ctx)` is we get both the replica context tags (`r1/1:...`) and the queue name the same as the logs. 

I recall we talked about there being an issue with the pprof labels not being clear
___
*[`pkg/kv/kvserver/testing_knobs.go` line 469 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NVjbWGW0ynw2Om25pyZ:-NVjbWGW0ynw2Om25py_:bfmx2gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dedabe181b7a7735aed2cd1e86144f8a50188bbd/pkg/kv/kvserver/testing_knobs.go#L469)):*
> ```Go
> 	RangeLeaseAcquireTimeoutOverride time.Duration
> 
> 	// MaybeAddInterceptor intercepts calls to MaybeAdd to verify label name is
> ```

nit: add period.
___
*[`pkg/kv/kvserver/testing_knobs.go` line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NVjbg__9LrXABojtRmb:-NVjbg__9LrXABojtRmc:bhsk9zw) ([raw file](https://github.com/cockroachdb/cockroach/blob/dedabe181b7a7735aed2cd1e86144f8a50188bbd/pkg/kv/kvserver/testing_knobs.go#L471)):*
> ```Go
> 	// MaybeAddInterceptor intercepts calls to MaybeAdd to verify label name is
> 	// set correctly during MaybeAdd
> 	MaybeAddInterceptor func(labelName string, labelDoesExist bool)
> ```

This interceptor is very specific to testing labeling, is there a way to make this more general? Would using the context and perhaps the replica be fine?
___
*[`pkg/kv/kvserver/queue_test.go` line 567 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NVjbwlBAD5ixNWSNwKq:-NVjbwlBAD5ixNWSNwKr:bds3i29) ([raw file](https://github.com/cockroachdb/cockroach/blob/dedabe181b7a7735aed2cd1e86144f8a50188bbd/pkg/kv/kvserver/queue_test.go#L567)):*
> ```Go
> // TestBaseQueueLabel verifies that queue name tag does not exist before and
> // after maybeAdd called shouldQueue but exists while shouldQueue was called.
> func TestBaseQueueLabel(t *testing.T) {
> ```

Nice test!


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: annotate queues with pprof labels

Prior to this commit, the debugging process with baseQueue functions was
complicated due to the difficulty in determining which specific queue invoked
the function of the baseQueue while examining the goroutine dump. This lack of
clarity is because different types of queues (consistency, merge, replica, etc.)
all utilize the baseQueue as their base implementation.

To resolve this issue, this commit introduces pprof labels to key baseQueue
functions like maybeAdd and processLoop. As a result, when examining the
goroutine dump, the pprof label (queueName, "") is displayed to indicate which
specific queue is stuck on the function call.

Fixes: https://github.com/cockroachdb/cockroach/issues/102222

wenyihu6 commented:
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NVjWmAw4CPb6IABzNPD):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Those queues still re-use the baseQueue as the base implementation for their queue after this commit. Perhaps try rewording the follow on sentence and using that after "Prior to this commit".
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/testing_knobs.go` line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NVjbg__9LrXABojtRmb:-NVo7o_50ix59JhC2VEh:b-4jypm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/dedabe181b7a7735aed2cd1e86144f8a50188bbd/pkg/kv/kvserver/testing_knobs.go#L471)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This interceptor is very specific to testing labeling, is there a way to make this more general? Would using the context and perhaps the replica be fine?
</blockquote></details>

Is it better now? I wasn't sure what you meant by `replica`. Are we referring to the `replica` in queue?


<!-- Sent from Reviewable.io -->

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-:-NW35y2I2TASvms-7-qs:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/queue_test.go` line 578 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103525#-NW367kADR3WlIRzVprI:-NW367kADR3WlIRzVprJ:b8b3df3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9362254d8a47d1d187c0e0051b553c4b4a67f6dc/pkg/kv/kvserver/queue_test.go#L578)):*
> ```Go
> 	r, err := tc.store.GetReplica(1)
> 	if err != nil {
> 		t.Fatal(err)
> ```

nit: use `require.NoError` instead here.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR! bors r=kvoli

wenyihu6 commented:
bors r=kvoli
